### PR TITLE
constFeature implementation for DAAL RF memorySavingMode

### PR DESCRIPTION
**Pull request by icfaust**
# Description
DAAL RF code currently only allows for constant features to not be used when memorySavingMode = False.  This uses the splitting functions themselves to find constant Features and add them to the constant feature list.

Changes proposed in this pull request:
- Adds constFeature bool to findBestSplitForFeature()
- Adds code for checking for constant data to CategoricalFeature and OrderedFeature in splitter classes in Regression and Classification
- Removes certain if statement memorySaveMode checks

### CODE REVIEW

File: cpp/daal/src/algorithms/dtrees/forest/classification/df_classification_train_dense_default_impl.i [hunk 1]
```diff
@@ -447,14 +452,15 @@ bool UnorderedRespHelperBest<algorithmFPType, cpu>::findSplitCategoricalFeature(
...
-                                                                                const algorithmFPType totalWeights) const
+                                                                                const algorithmFPType totalWeights, bool & constFeature) const
 {
     DAAL_ASSERT(n >= 2 * nMinSplitPart);
     _impRight.init(_nClasses);
     bool bFound                       = false;
     const bool bBestFromOtherFeatures = !(split.impurityDecrease < 0);
     algorithmFPType vBest             = -1;
     IndexType iBest                   = -1;
+    algorithmFPType first;
```
Comment (KulikovNikita) [inline cpp/daal/src/algorithms/dtrees/forest/classification/df_classification_train_dense_default_impl.i:463]:
```suggestion
    algorithmFPType first{};
```
Static analyzers don't like uninitialized variables

**Review by KulikovNikita — APPROVED** [review]

File: cpp/daal/src/algorithms/dtrees/forest/classification/df_classification_train_dense_default_impl.i (hunk 1 above)
Comment (icfaust) [reply cpp/daal/src/algorithms/dtrees/forest/classification/df_classification_train_dense_default_impl.i]: Thanks for the suggestion, made changes related to it.

**Review by ahuber21 — CHANGES_REQUESTED** [review]
Did you run the usual performance benchmarks? Are they unchanged?

File: cpp/daal/src/algorithms/dtrees/forest/classification/df_classification_train_dense_default_impl.i [hunk 2]
```diff
@@ -500,6 +506,9 @@ bool UnorderedRespHelperBest<algorithmFPType, cpu>::findSplitCategoricalFeature(
         split.featureValue = first;
         bFound             = true;
     }
+
+    constFeature = first == featureVal[0];
```
[the remark this answers is not in the document]
Comment (icfaust) [reply cpp/daal/src/algorithms/dtrees/forest/classification/df_classification_train_dense_default_impl.i:510]: _featHelper is assigned to a value definfed by const reference meaning that a common version is used by all trees.  We would need to duplicate this object for all trees and then modify the underlying class definition to contain the necessary level and const information.

The use of _aConstFeatureIdx collects data from a separate function (https://github.com/oneapi-src/oneDAL/blob/master/cpp/daal/src/algorithms/dtrees/dtrees_train_data_helper.i#L380, called by https://github.com/oneapi-src/oneDAL/blob/master/cpp/daal/src/algorithms/dtrees/forest/df_train_dense_default_impl.i#L1169 ). At worst case, hasDifferentFeatureValues unnecessarily loads almost the entire feature value vector a second time from memory just for a comparison. This accesses, like histogram, are not vectorizable (as far as I can tell) due to splitting and bootstrapping. This PR attempts to achieve results of finding constant nature of a feature without using IndexedFeatures while maintaining performance (i.e. no second data load).

**Comment by icfaust** [comment]
There seems to be an seg fault issue when running it in sklearn tests. I want to hold off on this until I have more time to find out possible causes. Until then, ExtraTrees will move to the hist method due to maturity in the underlying code.

**Pull request closed without merging**